apm-agent-core/src/main/java/co/elastic/apm/agent/bci/ElasticApmAgent.java

```diff
@@ -239,6 +239,12 @@ private static synchronized void initInstrumentation(final ElasticApmTracer trac
             adviceClassName2instrumentationClassLoader.put(
                 apmInstrumentation.getAdviceClassName(),
                 apmInstrumentation.getClass().getClassLoader());
+            if (apmInstrumentation instanceof TracerAwareInstrumentation) {
+                TracerAwareInstrumentation tracerAwareInstrumentation = (TracerAwareInstrumentation) apmInstrumentation;
+                pluginPackages2pluginClassLoaderRootPackages.putIfAbsent(tracerAwareInstrumentation.getPluginPackage(), new HashSet<String>());
+                Set<String> pluginClassLoaderPackages = pluginPackages2pluginClassLoaderRootPackages.get(tracerAwareInstrumentation.getPluginPackage());
+                pluginClassLoaderPackages.addAll(tracerAwareInstrumentation.pluginClassLoaderRootPackages());
```

Comment:
That was an explicit design decision. I'll document it in the JavaDocs of `co.elastic.apm.agent.bci.TracerAwareInstrumentation#pluginClassLoaderRootPackages`.

The issue is that while regular instrumentations are loaded on startup, dynamic instrumentations may or may not be applied at a later point in time. However, it is important that the set of plugin root packages is known upfront and doesn't change. That's because IndyPluginClassLoaders might have already been created before `DynamicTransformer.ensureInstrumented` has been called.